File: nbodyf/nbody.f90
program SimulateNBody

    implicit none

    type Nbody
        integer :: id
        double precision, dimension(3) :: x
        double precision, dimension(3) :: vel
        double precision, dimension(3) :: force
        double precision :: mass
        double precision :: potentialE
    end type Nbody    

    real :: display = 1.3
    integer :: npts=1500
    integer :: nsteps=500
    integer :: outevery = 5
    integer :: simulation = 1
    integer :: output=1
    integer :: i
    integer :: seed = 5
    double precision :: dt = 0.01
    double precision :: time = 0.
    double precision :: tote = 0.

    type(Nbody), allocatable :: pdata(:)

    allocate(pdata(npts))

    call random_seed(seed)

    call initialize_particles(pdata, npts, simulation)
    call calculate_forces_fastest(pdata, npts)
    call calculate_energy(pdata, npts, tote)

    do i=1,nsteps
        call nbody_step(pdata, npts, dt)
        call calculate_energy(pdata, npts, tote)
        time = time + dt
        if (output /= 0) then
            print *, i, dt, time, tote
            if (mod(i,outevery) == 0) then 
                call display_particles(pdata, npts, display)
            endif
        endif
    enddo

    contains

    subroutine initialize_particles(pdata, n, simulation)
        implicit none
        integer, intent(IN) :: n
        type(Nbody), intent(INOUT), dimension(n) :: pdata
        integer, intent(IN) :: simulation

        integer :: i, j
        double precision, parameter :: vamp = 0.1
        double precision :: r
        double precision :: rnd
        double precision ::  mass

        mass = 1./(1.*n)

        if (simulation == 1) then
            i = 1
            do while (i<=n)
                r = 0.
                do j=1,3
                    call random_number(rnd)
                    pdata(i) % x(j) = 2.0 * rnd - 1.
                    r = r + pdata(i) % x(j)**2
                enddo
                if (r < 1.) then
                    pdata(i) % mass = mass
                    do j=1,3
                        call random_number(rnd)
                        pdata(i) % vel(j) = vamp * ( 2.0 * rnd - 1.)
                    enddo
                    pdata(i) % id = i
                    i = i + 1
                endif
            enddo
        endif
    return
    end subroutine initialize_particles


    subroutine calculate_forces_fastest(pdata, n)
        implicit none
        integer, intent(IN) :: n
        type(Nbody), intent(INOUT), dimension(n) :: pdata
        integer :: i, j, d
        double precision :: rsq
        double precision, parameter :: EPS=0.1
        double precision, parameter :: gravconst=1.
        double precision, dimension(3) :: dx
        double precision :: ir
        double precision :: forcex

        do i=1,n
            pdata(i) % force = 0.
            pdata(i) % potentialE = 0.
        enddo

        do i=1,n
            do j=i+1,n
                rsq = EPS*EPS
                dx = 0.
                do d=1,3
                    dx(d) = pdata(j)%x(d) - pdata(i)%x(d)
                    rsq = rsq + dx(d)*dx(d)
                enddo
                ir = 1./sqrt(rsq)
                rsq = ir/rsq
                do d=1,3
                    forcex = rsq*dx(d) * pdata(i)%mass * pdata(j)%mass * gravconst
                    pdata(i)%force(d) = pdata(i)%force(d) + forcex
                    pdata(j)%force(d) = pdata(j)%force(d) - forcex
                enddo
                pdata(i)%potentialE = pdata(i)%potentialE -             &
                     gravconst * pdata(i)%mass * pdata(j)%mass * ir  
                pdata(j)%potentialE = pdata(i)%potentialE -             &
                     gravconst * pdata(i)%mass * pdata(j)%mass * ir
            enddo
        enddo

        return
    end subroutine calculate_forces_fastest

    subroutine calculate_energy(pdata, n, energy)
        implicit none
        integer, intent(IN) :: n
        type(NBody), intent(INOUT), dimension(n) :: pdata
        double precision, intent(OUT) :: energy
        
        integer :: i

        energy = 0.
        do i=1,n
           energy = energy + pdata(i)%potentialE  + &
            0.5 * sum(pdata(i)%vel*pdata(i)%vel) * pdata(i)%mass
        enddo

        return
    end subroutine calculate_energy



    subroutine nbody_step(pdata, n, dt)
        implicit none
        integer, intent(IN) :: n
        double precision, intent(IN) :: dt
        type(NBody), intent(INOUT), dimension(n) :: pdata

        integer :: i
        double precision, dimension(3) :: accel

        ! kick
        do i=1,n
            pdata(i)%x = pdata(i)%x + pdata(i)%vel*dt
        enddo

        ! drift
        call calculate_forces_fastest(pdata, n)

        do i=1,n
            accel = pdata(i)%force / pdata(i)%mass
            pdata(i)%vel = pdata(i)%vel + accel*dt
        enddo
       
        return
    end subroutine nbody_step


    subroutine display_particles(pdata, n, cmax)
        implicit none
        integer, intent(IN) :: n
        type(NBody), intent(IN), dimension(n) :: pdata
        real, intent(IN) :: cmax

        integer, parameter :: npix=256
        real, dimension(6) :: tr
        real, dimension(npix,npix) :: plot
        logical, save :: firsttime = .true.

        integer :: i, ii, jj

        if (firsttime) then
            call pgbeg(0, "/xwindow", 1, 1)
            call pgask(0)
            call pgenv(1.,npix*1.,1.,npix*1.,1,2)
            call pglab('x','y','View from top')
            firsttime = .false.
        endif

        tr = (/ 0,1,0,0,0,1 /)
        plot = 0.

        do i=1,n
            ii = npix*int((pdata(i)%x(1) + cmax)/(2.*cmax))
            jj = npix*int((pdata(i)%x(2) + cmax)/(2.*cmax))
            if ( (ii > 0) .and. (ii <= npix) .and. (jj > 0) .and. (jj <= npix) ) then
                plot(ii,jj) = plot(ii,jj) - 1.
            endif
        enddo

        call pgimag(plot,npix,npix,1,npix,1,npix,-5.0,.0,tr)
        return
    end subroutine display_particles  

end program 

File: nbodyc/nbody.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <assert.h>
#include <omp.h>
#include <getopt.h>
#include "allocmultid.h"

#ifdef PGPLOT
#include <cpgplot.h>
#endif

#define NDIM 3
#define EPS 0.1

#define GRAVCONST 1.         /* sets timescale */
#define SIM_RANDOM  1        /* which ICs are we going to use? */

/*--------------------------------------------------------------------------------*/


typedef struct nbody_struct_s {
      double x[NDIM];    /*the particle positions*/
      double v[NDIM];    /*the particle velocities*/
      double f[NDIM];    /*the forces on the particles*/
      double mass;
      double PE;        /* potential energy */
} NBody;


/*--------------------------------------------------------------------------------*/

NBody *allocate_nbody(int n)
{
   NBody *mydata;
   mydata=(NBody *)malloc(n * sizeof(NBody));
   return mydata;
}

/*--------------------------------------------------------------------------------*/

void initialize_particles(NBody *data, int n, int simulation) 
{
   int i;
   double vamp = 0.1; /* KE/U ~ 5/6 vamp^2 */

   switch (simulation) {
      case SIM_RANDOM:
         i = 0;
         while (i < n) {
            double r = 0.;
            for (int j=0; j<NDIM; j++) {
               data[i].x[j] = 2.0*rand()/RAND_MAX - 1;
               r += data[i].x[j] * data[i].x[j];
          
               data[i].v[j]=vamp*(2.0*rand()/RAND_MAX - 1.);
               data[i].f[j]=0;
            } 
            if (r > 1.) continue;

            data[i].mass = 1./n;
            i++;
         }
         break;

   } 
   return;
}


/*--------------------------------------------------------------------------------*/

void calculate_forces_fastest(NBody *data, int n)
{
   for (int i=0;i<n;i++) {
      for (int j=0;j<NDIM;j++) {
         data[i].f[j]=0; 
      }
      data[i].PE = 0.;
   }

   for (int i=0;i<n;i++){
      for (int j=0;j<i;j++) {
         double rsq=EPS*EPS, dx[NDIM],forcex;
         for (int k=0;k<NDIM;k++){
            dx[k]=data[j].x[k]-data[i].x[k];
            rsq+=dx[k]*dx[k];
         }
         double ir =1./sqrt(rsq);
         rsq=ir/rsq;
         for (int k=0;k<NDIM;k++)
         {
            forcex=rsq*dx[k] * data[i].mass * data[j].mass * GRAVCONST;
            data[i].f[k] += forcex;
            data[j].f[k] -= forcex;
         }
         data[i].PE -= GRAVCONST * data[i].mass * data[j].mass * ir;
         data[j].PE -= GRAVCONST * data[i].mass * data[j].mass * ir;
      }
   }
}

/*--------------------------------------------------------------------------------*/

void calculate_energy(NBody *data, int n, double  *energy) {
   *energy = 0.;
  
   for (int i=0; i<n; i++) {
      *energy += 0.5*data[i].PE;
      double ke = 0.; 
      for (int j=0; j<NDIM; j++) 
         ke += (data[i].v[j])*(data[i].v[j]);
      ke *= 0.5 * data[i].mass;
      *energy += ke;
   }

   return;
}

/*--------------------------------------------------------------------------------*/

void calculate_forces(NBody *data, int n)
{
   calculate_forces_fastest(data,n);
   //calculate_forces_fastest_omp(data);
				  
   return;
}

/*--------------------------------------------------------------------------------*/

/* kick-drift-kick */
void nbody_step(NBody *data, int n, double dt)
{
   for (int i=0;i<n;i++)
      for (int j=0;j<NDIM;j++) {
         data[i].x[j] += data[i].v[j]*dt;
      }
   calculate_forces(data, n);
   for (int i=0;i<n;i++)
      for (int j=0;j<NDIM;j++) {
         double a = data[i].f[j] / data[i].mass;
         data[i].v[j] += a*dt;
      }

   return;
}

/*--------------------------------------------------------------------------------*/

void display_particles(NBody *data, double max, int n)
/*assume a top-down view (i.e. ignore the z-component*/
{
#ifdef PGPLOT
   int npix=256;
   float tr[6];
   static int firsttime=1;

   if (firsttime)
   {
      assert(NDIM>=2);
      cpgopen("?");
      cpgenv(1,npix,1,npix,0,1);
      firsttime=0;
   }
   tr[0]=0;tr[1]=1;tr[2]=0;tr[3]=0;tr[4]=0;tr[5]=1;
   float **mat=alloc2d_float(npix,npix);
   for (int i=0;i<npix;i++)
      for (int j=0;j<npix;j++)
         mat[i][j]=0;
  
   for (int i=0;i<n;i++)
   {
      int ii= npix*((data[i].x[0]+max)/(2*max));
      int jj= npix*((data[i].x[1]+max)/(2*max));
      if ((ii>=0)&&(ii<npix)&&(jj>=0)&&(jj<npix))
         mat[ii][jj]++;
   }

   cpgimag(mat[0],npix,npix,1,npix,1,npix,0,5.0,tr);
#endif
  
}


/*--------------------------------------------------------------------------------*/

int get_options(int argc, char **argv, int *npts, int *nsteps, int *dooutput, int *outevery, int *simulation, double *dt) {

   int output = *dooutput;

   const struct option long_options[] = {
      {"output"   , no_argument, &output, 1},
      {"nooutput" , no_argument, &output, 0},
      {"npts"     , required_argument, 0, 'n'},
      {"nsteps"   , required_argument, 0, 't'},
      {"simulation", required_argument, 0, 's'},
      {"outevery", required_argument, 0, 'o'},
      {"dt", required_argument, 0, 'd'},
      {"nproc", required_argument, 0, 'P'},
      {"help",      no_argument, 0, 'h'},
      {0, 0, 0, 0}};

   char c;
   int option_index;
   int tempint;
   double tempflt;

   while (1) {
      c = getopt_long(argc, argv, "n:t:s:o:P:h:e", long_options, &option_index);
      if (c == -1) break;

      switch (c) {
         case 0: if (long_options[option_index].flag != 0)
            break;

         case 'n': tempint = atoi(optarg);
            if (tempint < 1 || tempint > 100000) {
               fprintf(stderr,"%s: Cannot use number of points %s;\n  Using %d\n", argv[0], optarg, *npts);
            } else {
               *npts = tempint;
            }
            break;


         case 't': tempint = atoi(optarg);
            if (tempint < 1 ) {
               fprintf(stderr,"%s: Cannot use number of steps  %s;\n  Using %d\n", argv[0], optarg, *nsteps);
            } else {
               *nsteps  = tempint;
            }
            break;

         case 'o': tempint = atoi(optarg);
            if (tempint < 1) {
               fprintf(stderr,"%s: invalid output frequency %s;\n  Using %d\n", argv[0], optarg, *outevery);
            } else {
               *outevery = tempint;
            }
            break;

         case 's': tempint = atoi(optarg);
            if (tempint != SIM_RANDOM) {
               fprintf(stderr,"%s: invalid simulation number %s;\n  Using %d\n", argv[0], optarg, *simulation);
            } else {
               *simulation = SIM_RANDOM;
            }
            break;


         case 'd': tempflt = atof(optarg);
            if (tempflt <= 0. || tempflt > 1.) {
               fprintf(stderr,"%s: invalid timestep parameter %s;\n  Using %g\n", argv[0], optarg, *dt);
            } else {
               *dt = tempflt;
            }
            break;

         default: printf("Invalid option %s\n", optarg);

         case 'h':
            puts("Options: ");
            puts("    --npts=N      (-n N): Set the number of particles.");
            puts("    --nsteps=N    (-t N): Set the number of time steps taken.");
            puts("    --simulation=N(-s N): Set the simulation; 0 == Random ICs.");
            puts("    --outevery=N  (-o N): Set output every N timesteps.");
            puts("    --dt=X  (-d X): Timestep size dt.");
            puts("");
            return +1;

            break;
      }
   }

   *dooutput = output;
   return 0;
}


/*================================================================================*/

int main(int argc, char *argv[])
{
   int npts = 1500;
   int nsteps = 500;
   int outevery = 1;
   int simulation = SIM_RANDOM;
   int status;
   int output = 1;
   double dt = 0.01;
   double time = 0.;
   double tote = 0.;

   status = get_options(argc, argv, &npts, &nsteps, &output, &outevery, &simulation, &dt);
   if (status) return 0;

   NBody *mydata;


   mydata=allocate_nbody(npts);
   printf("mydata has %d particles in a %d-dimensional space.\n",npts,NDIM);
  
   initialize_particles(mydata, npts, simulation);
  
   calculate_forces(mydata, npts);
   calculate_energy(mydata, npts, &tote);

   for (int i=0;i<nsteps;i++)
   {
      nbody_step(mydata,npts,dt);
      calculate_energy(mydata, npts, &tote);
      time += dt;
      if (output) {
         printf("%i\t%g\t%g\t%g\n", i, dt, time, tote);
         if (!(i % outevery)) display_particles(mydata,1.3,npts);
      }
   }
  
  
   FILE *outfile=fopen("myparticles.txt","w");
   fprintf(outfile,"positions and forces are:\n");
   for (int i=0;i<npts;i++) 
   {
      for (int j=0;j<NDIM;j++)
         fprintf(outfile,"%12.6f ",mydata[i].x[j]);
      fprintf(outfile,"      ");
      for (int j=0;j<NDIM;j++)
         fprintf(outfile,"%12.6f ",mydata[i].f[j]);
      fprintf(outfile,"\n");
   }


   return 0;
}
